com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java

```diff
@@ -0,0 +1,110 @@
+        StackFrameReference stackFrameReference = (StackFrameReference) context.getRecyclableIdPool().getObjectById(restartFrameArgs.frameId);
+
+        if (stackFrameReference == null) {
+            return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.RESTARTFRAME_FAILURE,
+                    String.format("RestartFarme: cannot find the stack frame with frameID %s", restartFrameArgs.frameId));
+        }
+
+        if (canRestartFrame(context, stackFrameReference)) {
+            try {
+                ThreadReference reference = stackFrameReference.getThread();
+                popStackFrames(context, reference, stackFrameReference.getDepth());
+                stepin(context, reference);
+            } catch (DebugException de) {
+                context.getProtocolServer().sendEvent(new Events.UserNotificationEvent(NotificationType.ERROR, de.getMessage()));
+            }
+            return CompletableFuture.completedFuture(response);
+        } else {
+            context.getProtocolServer().sendEvent(new Events.UserNotificationEvent(NotificationType.ERROR, "Current stack frame doesn't support restart."));
+            context.getProtocolServer().sendEvent(new Events.StoppedEvent("restartframe", stackFrameReference.getThread().uniqueID()));
+            return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.RESTARTFRAME_FAILURE, "Failed to restart the selected stack frame.");
+        }
+    }
+
+    private boolean canRestartFrame(IDebugAdapterContext context, StackFrameReference frameReference) {
+        ThreadReference reference = frameReference.getThread();
+        StackFrame[] frames = context.getStackFrameManager().reloadStackFrames(reference);
+        // Current stack frame cannot be the top frame of the thread:
+        if (frames.length <= frameReference.getDepth() + 1) {
+            return false;
+        }
+
+        // Cannot restart frame involved with native call stacks:
+        for (int i = 0; i <= frameReference.getDepth() + 1; i++) {
+            if (StackFrameUtility.isNative(frames[i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void popStackFrames(IDebugAdapterContext context, ThreadReference thread, int depth) throws DebugException {
+        while (depth >= 0) {
+            StackFrame[] frames = context.getStackFrameManager().reloadStackFrames(thread);
+            StackFrameUtility.pop(frames[0]);
+            --depth;
+        }
+    }
+
+    private void stepin(IDebugAdapterContext context, ThreadReference thread) {
+        StepRequest request = DebugUtility.createStepIntoRequest(thread, context.getStepFilters().classNameFilters);
+        context.getDebugSession().getEventHub().stepEvents().filter(debugEvent -> request.equals(debugEvent.event.request())).take(1).subscribe(debugEvent -> {
+            debugEvent.shouldResume = false;
+            // Have to send to events to keep the UI sync with the step in operations:
```

Comment:
typos `to` -> `two`.

Reply:
`with the step in operations` is a little confusing at the quick glance, how about changing it to `with the latest stepIn status`.